File: frontend/src/api/leadershipTemplates.ts
import { apiBase, apiRequest } from "../lib/api/client";

export type LeadershipTemplate = { id: string; name: string; role: string; document_type: "LEADERSHIP_RECOGNITION" | "END_OF_TENURE_APPRECIATION"; signature_handling: "retain" | "replace" | null; active: boolean; archived: boolean };
export type LeadershipField = { field_name: string; page_number: number; x: number; y: number; width: number; height: number };
export type ExecutiveMembership = { id: string; student_id: string; session_id: string; society_id: string | null; role: string; start_date: string; end_date: string | null; status: "ACTIVE" | "COMPLETED" | "REMOVED" };
export const listLeadershipTemplates = () => apiRequest<LeadershipTemplate[]>("/admin/leadership-templates");
export const listExecutiveMemberships = () => apiRequest<ExecutiveMembership[]>("/admin/executive-memberships");
export const activateLeadershipTemplate = (id: string) => apiRequest<LeadershipTemplate>(`/admin/leadership-templates/${id}/activate`, { method: "POST" });
export const deactivateLeadershipTemplate = (id: string) => apiRequest<LeadershipTemplate>(`/admin/leadership-templates/${id}/deactivate`, { method: "POST" });
export const archiveLeadershipTemplate = (id: string) => apiRequest<LeadershipTemplate>(`/admin/leadership-templates/${id}/archive`, { method: "POST" });
export const configureLeadershipTemplate = (id: string, fields: LeadershipField[], signature_handling: "retain" | "replace") => apiRequest<LeadershipTemplate>(`/admin/leadership-templates/${id}/fields`, { method: "POST", body: JSON.stringify({ fields, signature_handling }) });
export async function uploadLeadershipTemplate(name: string, role: string, document_type: LeadershipTemplate["document_type"], file: File) { const body = new FormData(); body.set("name", name); body.set("role", role); body.set("document_type", document_type); body.set("file", file); return apiRequest<LeadershipTemplate>("/admin/leadership-templates/upload", { method: "POST", body }); }
export async function previewLeadershipTemplate(id: string, membership_id: string): Promise<Blob> { const response = await fetch(`${apiBase}/admin/leadership-templates/${id}/preview`, { method: "POST", credentials: "include", headers: { "Content-Type": "application/json" }, body: JSON.stringify({ membership_id }) }); if (!response.ok) { const body = await response.json().catch(() => null) as { detail?: string } | null; throw new Error(body?.detail ?? "The preview could not be generated."); } return response.blob(); }

File: backend/app/schemas/leadership_templates.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field

LeadershipDocumentType = Literal["LEADERSHIP_RECOGNITION", "END_OF_TENURE_APPRECIATION"]


class LeadershipTemplateFieldInput(BaseModel):
    field_name: str = Field(pattern=r"^[a-z][a-z0-9_]{0,79}$")
    page_number: int = Field(ge=1)
    x: int = Field(ge=0)
    y: int = Field(ge=0)
    width: int = Field(gt=0)
    height: int = Field(gt=0)


class LeadershipTemplateFieldsCreate(BaseModel):
    fields: list[LeadershipTemplateFieldInput] = Field(min_length=1)


class LeadershipTemplateResponse(BaseModel):
    id: UUID
    name: str
    role: str
    document_type: LeadershipDocumentType
    active: bool
    archived: bool
    created_at: datetime
